pytheus/backends.py

```diff
@@ -1,10 +1,92 @@
+from abc import ABC, abstractmethod
+import importlib
+import json
+import os
 from threading import Lock
+from typing import Any, Optional, Tuple
 
+from pytheus.exceptions import InvalidBackendClassException, InvalidBackendConfigException
 
-class LockedValue:
-    """Provides a value protected by a threading Lock."""
+BackendConfig = dict[str, Any]
 
-    def __init__(self) -> None:
+
+class Backend(ABC):
+    def __init__(self, config: BackendConfig):
+        self.config = config  # TODO: We should discuss validation here (jsonschema, pydantic, etc)
+
+    @abstractmethod
+    def inc(self, value: float) -> None:
+        pass
+
+    @abstractmethod
+    def get(self) -> float:
+        pass
+
+
+def _import_backend_class(full_import_path: str) -> Backend:
+    try:
+        module_path, class_name = full_import_path.rsplit(".", 1)
+    except ValueError:  # Empty string or not full path to the class
+        raise InvalidBackendClassException(
+            "Backend class could not be imported. Full import path needs to be provided, e.g. "
+            "my_package.my_module.MyClass"
+        )
+    try:
+        module = importlib.import_module(module_path)
+    except ImportError as e:
+        raise InvalidBackendClassException(f"Module '{module_path}' could not be imported: {e}")
+    try:
+        cls = getattr(module, class_name)
+        if not issubclass(cls, Backend):
+            raise InvalidBackendClassException(f"Class '{class_name}' is not a Backend subclass")
+        return cls
+    except AttributeError:
+        raise InvalidBackendClassException(
+            f"Class '{class_name}' could not be found in module '{module_path}'"
+        )
+
+
+def load_backend(
+    backend_class: Optional[Backend] = None,
+    backend_config: Optional[BackendConfig] = None,
+):
+
+    # Load default backend class
+    global BACKEND_CLASS
+    backend_class_env_var: str = "PYTHEUS_BACKEND_CLASS"
+    if backend_class is not None:  # Explicit
+        BACKEND_CLASS = backend_class
+    elif backend_class_env_var in os.environ:  # Environment
+        class_name = os.environ[backend_class_env_var]
+        BACKEND_CLASS = _import_backend_class(os.environ[backend_class_env_var])
+    else:  # Default
+        BACKEND_CLASS = SingleProcessBackend
+
+    # Load default backend config
+    global BACKEND_CONFIG
+    backend_config_env_var: str = "PYTHEUS_BACKEND_CONFIG"
+    if backend_config is not None:  # Explicit
+        BACKEND_CONFIG = backend_config
+    elif backend_config_env_var in os.environ:  # Environment
+        try:
+            with open(os.environ[backend_config_env_var]) as f:
+                BACKEND_CONFIG = json.loads(f.read())  # TODO: Support yaml?
+        except Exception as e:
+            raise InvalidBackendConfigException(f"{e.__class__}: {e}")
+    else:
+        BACKEND_CONFIG = {}  # Default
+
+
+def get_backend() -> Backend:
+    # Probably ok not to cache this and allow each metric to keep its own
```

Comment:
just a note, on this we might improve it in the future for performance reasons, but for now it's absolutely ok

Reply:
Yeah one reason not to cache it could be that you swap backends in the middle of the code and want the new metrics to get the new one.

But yeah, this operation is minimal in terms of CPU/memory, and we won't have more than as many `MetricCollectors` (We are talking here about a couple of hundreds, at most?)